File: CppStack/src/stack.cpp
#ifndef STACK_CPP
#define STACK_CPP

#include "stack.h"
#include "node.h"
#include <cstddef>

template <typename T>
Stack<T>::~Stack()
{
    Node<T> *current = top;

    while (top != nullptr)
    {
        Node<T> *next_node = current->next;
        delete current;
        current = next_node;
    }
};

template <typename T>
void Stack<T>::push(const T value)
{
    Node<T> *new_node = new Node<T>(value);

    if (top != nullptr)
        new_node->next = top;

    top = new_node;
}

template <typename T>
T Stack<T>::pop()
{
    if (!is_empty())
    {
        T value = top->value;
        top = top->next;
        return value;
    }

    return T();
}

template <typename T>
bool Stack<T>::is_empty() const
{
    return top == nullptr;
}

template <typename T>
bool Stack<T>::query(const T value, Node<T> **currentNode, Node<T> **previousNode) const
{
    *currentNode = start;
    *previousNode = nullptr;

    while (*currentNode != nullptr)
    {
        if ((*currentNode)->value == value)
        {
            return true;
        }

        *previousNode = *currentNode;
        *currentNode = (*currentNode)->next;
    }

    return false;
}

template <typename T>
T *Stack<T>::to_array(size_t &size) const
{
    size = 0;
    Node<T> *current = top;

    while (current != nullptr)
    {
        size++;
        current = current->next;
    }

    T *values = new T[size];
    current = top;
    int index = 0;

    while (current != nullptr)
    {
        values[index] = current->value;
        index++;
        current = current->next;
    }

    return values;
}

#endif

File: CsharpStack/Stack.cs
namespace CsharpStack;

public class MyStack<T>
{
    private Node<T>? Top { get; set; }
    private static Node<T> GetNode(T value) => new Node<T>(value);

    public MyStack()
    {
        Top = null;
    }

    public bool IsEmpty()
    {
        return Top == null;
    }

    public T? GetTop()
    {
        return !IsEmpty() ? Top!.Value : default;
    }

    public void Push(T value)
    {
        var newNode = GetNode(value);
        if (!IsEmpty()) newNode.Next = Top;
        Top = newNode;
    }

    public T? Pop()
    {
        if (!IsEmpty())
        {
            var value = Top!.Value;
            Top = Top.Next;
            return value;
        }

        return default;
    }

    public bool Query(T value, out Node<T>? currentNode, out Node<T>? previousNode)
    {
        currentNode = Top;
        previousNode = null;

        while (currentNode != null)
        {
            if (currentNode.Value!.Equals(value))
                return true;

            previousNode = currentNode;
            currentNode = currentNode.Next;
        }

        currentNode = null;
        previousNode = null;

        return false;
    }

    public override string ToString()
    {
        if (IsEmpty()) return "[]";

        Node<T>? aux = Top;

        string result = "";

        while (aux != null)
        {
            var isLastNode = aux.Next == null;
            result += string.Format("{0}{1}", aux.Value, !isLastNode ? " -> " : "");
            aux = aux.Next;
        }

        return string.Format("[START -> {0} -> END]", result);
    }
}
